pkg/gadgets/trace/signal/types/types.go

```diff
@@ -15,18 +15,25 @@
 package types
 
 import (
+	"github.com/kinvolk/inspektor-gadget/pkg/columns"
```

Comment:
I talked to @blanquicet yesterday especially about this and I would like to have the columns definition inside the types.go of the corresponding package: Whenever we need special handling of columns (like custom extractors or virtual columns), this way we can also have that definition in the same file. So you can be sure to have the Columns on every gadget and it will give you information about the metadata available. Also, the library will probably be used by not only the frontend, but also by the gadgets themselves (e.g. sorting, filtering, grouping on the server side before cutting the result set to a limited amount like in the top gadgets).

Reply:
Ok, let's keep going this way. Later on we can split out the types if need to do so.

Reply:
> Also, the library will probably be used by not only the frontend, but also by the gadgets themselves (e.g. sorting, filtering, grouping on the server side before cutting the result set to a limited amount like in the top gadgets).

I was thinking about this better. I'm not fully convinced to use a library based on reflection in the backend to perform such a tasks. I agree that the library is very flexible to use in the frontend, but I'd like to avoid to use in the backend as much as possible because of performance reasons.

Reply:
I understand the concern of handling things via reflection and I see it more as a last resort for when the user wants to do specific things we didn't optimize for. That way the user can still do all the fancy things, even if it comes at some cost  - and for those additional features (that only cost when used), we don't have to write additional (struct specific) code at all.

Let's start with some benchmarks to see that reflection isn't all that bad (at least for our use cases):
```
BenchmarkFilterNumberGte-10        	49086614	        23.80 ns/op	       8 B/op	       1 allocs/op
BenchmarkFilterNumberEq-10         	50282662	        23.28 ns/op	       8 B/op	       1 allocs/op
BenchmarkFilterNumberNeq-10        	51283602	        23.31 ns/op	       8 B/op	       1 allocs/op
BenchmarkFilterStringEq-10         	34612514	        33.03 ns/op	      16 B/op	       1 allocs/op
BenchmarkFilterStringNeq-10        	35924289	        33.17 ns/op	      16 B/op	       1 allocs/op
BenchmarkFilterStringRegex-10      	31654264	        37.61 ns/op	       0 B/op	       0 allocs/op
BenchmarkFilterStringRegexCI-10    	17073394	        69.74 ns/op	       0 B/op	       0 allocs/op
BenchmarkGroup1000to100-10         	    6188	    189329 ns/op	  102683 B/op	     712 allocs/op
BenchmarkGroup1000to10-10          	    8006	    148145 ns/op	   63567 B/op	     105 allocs/op
BenchmarkSortReflection1000-10     	    3033	    393848 ns/op	     168 B/op	       4 allocs/op
BenchmarkSortNative1000-10         	    8763	    133296 ns/op	      56 B/op	       2 allocs/op
BenchmarkJSON-10                   	 3154509	       381.1 ns/op	     208 B/op	       1 allocs/op
```

These benchmarks are written for this struct:
```golang
type TestStruct struct {
	Int      int    `column:"int" json:"int"`
	Int2     int    `column:"int2" json:"int2"`
	Int3     int    `column:"int3" json:"int3"`
	Int4     int    `column:"int4" json:"int4"`
	String   string `column:"string" json:"string"`
	String2  string `column:"string2" json:"string2"`
	String3  string `column:"string3" json:"string3"`
	String4  string `column:"string4" json:"string4"`
	GroupID1 string `column:"groupID1" json:"groupID1"`
	GroupID2 string `column:"groupID2" json:"groupID2"`
}
```

(Source: https://github.com/flyth/inspektor-gadget/blob/michael/integrate-benchmark-tableformatter/pkg/columns/examples/benchmark/benchmark_test.go)

Now applying that to our use cases:

1. Heavy throughput gadgets sending events (e.g. trace gadgets)

The only real "feature" that is applicable here, is filtering. We already do filtering on pods, namespaces etc in a heavily optimized way for most gadgets - we should of course not change that. However, for all fields that are considered neither in BPF nor in native code, we could apply filtering through the columns lib (or even for fields like `pod` where we the optimized version support only matching for a specific name but not excluding a name / doing regex etc.)

The cost for a simple filter like `field >= int` in my tests was at about 23ns. If such a filter filters out an entry, we save about 390ns (for the rather simple struct I've tested with) for the JSON encoding and additional time and traffic for decoding / transferring. Given that the user chooses the filter wisely, I think this can help a lot actually decreasing the load on the nodes.

Regular expressions are of course slower, but that's where the overhead of using reflection becomes even smaller.

2. Snapshots (e.g. top gadgets)

For these, we need to do the same operations on the backend that we do on the frontend, as we only transfer a limited number of rows back to the client. So this is where we would need to handle grouping, sorting etc. This would happen every second (or less frequent) for a variable number of entries. The group/sort benchmark above handled 1000 pseudorandomly sorted entries. Those took ~0.4ms for sorting using reflection vs. ~0.13ms when doing it natively. Grouping 1000 entries down to 100 or 10 entries takes 0.19ms and 0.15ms - I have not written a native replacement for that.

Maybe we would need to do more testing with actual data, but I don't think using reflection looks that bad.

Reply:
I got your point. Yes, I agree that if the performance impact is not that big we can use this library on the backend to avoid having to write a lot of code for implementing some features. If we find that the reflection approach is too heavy we can always switch to a "native" implementation.